### PR TITLE
Quaternion-based Mahony's DCM IMU (initial implementation by S. Madgwick)

### PR DESCRIPTION
**Pull request by digitalentity**
This PR is for merging quaternion-based IMU from my navigation-rewrite. It is flight tested on CC3D and NAZE and it gives better attitude estimation and a more locked-in feeling in ANGLE and HORIZON modes.

### CODE REVIEW

**Comment by digitalentity** [comment]
Binaries for testing avaible here: https://www.dropbox.com/sh/4r5374mos7471q2/AAD4ah-K7vfA5cIDUgSXOtAba/PR1411?dl=0

**Comment by digitalentity** [comment]
This PR is merged to betaflight for testing, everything seems fine so far.

**Comment by hydra** [comment]
@digitalentity unit tests are failing, can you fix?

**Comment by digitalentity** [comment]
Strange, tests pass on my dev PC. Will doublecheck this, thanks

**Comment by digitalentity** [comment]
@hydra all fixed and rebased against latest master.

**Comment by digitalentity** [comment]
@hydra please don't merge this yet. There are a few changes being testing in betaflight that would improve this PR.

**Comment by digitalentity** [comment]
All fixes present. This is safe for merging now. Some more improvements are on the way, but they are too taxing on flash and don't improve performance for multicopters.

**Comment by borisbstyle** [comment]
@digitalentity 
Is this the same code as you did for betaflight?

**Comment by digitalentity** [comment]
Mostly, yes.

**Comment by digitalentity** [comment]
@hydra, I've rebased this PR. It can be cleanly merged now.

File: src/main/flight/imu.c [hunk 1]
```diff
@@ -53,22 +63,49 @@ uint32_t accTimeSum = 0;        // keep track for integration of acc
...
 static accDeadband_t *accDeadband;
 
+STATIC_UNIT_TESTED float q0 = 1.0f, q1 = 0.0f, q2 = 0.0f, q3 = 0.0f;    // quaternion of sensor frame relative to earth frame
+static float rMat[3][3];
+
+attitudeEulerAngles_t attitude = { { 0, 0, 0 } };     // absolute angle inclination in multiple of 0.1 degree    180 deg = 1800
+
+static float gyroScale;
+
+STATIC_UNIT_TESTED void imuCompureRotationMatrix(void)
```
Comment (lostcontrol) [inline src/main/flight/imu.c]: Likely a typo here: imuCompu**t**eRotationMatrix()

Comment (digitalentity) [inline src/main/flight/imu.c]: Thanks, fixing.

File: src/main/flight/imu.c [hunk 2]
```diff
@@ -174,126 +204,222 @@ void imuCalculateAcceleration(uint32_t deltaT)
...
+        while (yawError < -M_PIf) yawError += (2.0f * M_PIf);
+
+        ez += sin_approx(yawError / 2.0f);
+    }
 
-    // Apply complimentary filter (Gyro drift correction)
-    // If accel magnitude >1.15G or <0.85G and ACC vector outside of the limit range => we neutralize the effect of accelerometers in the angle estimation.
-    // To do that, we just skip filter, as EstV already rotated by Gyro
+    // Use measured magnetic field vector
+    recipNorm = mx * mx + my * my + mz * mz;
```
Comment (lostcontrol) [inline src/main/flight/imu.c]: I guess sq() could be used here (and everywhere square is computed) like on line 239

File: src/main/flight/imu.c [hunk 3]
```diff
@@ -174,126 +204,222 @@ void imuCalculateAcceleration(uint32_t deltaT)
...
+        mz *= recipNorm;
+
+        // For magnetometer correction we make an assumption that magnetic field is perpendicular to gravity (ignore Z-component in EF).
+        // This way magnetic field will only affect heading and wont mess roll/pitch angles
+
+        // (hx; hy; 0) - measured mag field vector in EF (assuming Z-component is zero)
+        // (bx; 0; 0) - reference mag field vector heading due North in EF (assuming Z-component is zero)
+        hx = rMat[0][0] * mx + rMat[0][1] * my + rMat[0][2] * mz;
+        hy = rMat[1][0] * mx + rMat[1][1] * my + rMat[1][2] * mz;
+        bx = sqrtf(hx * hx + hy * hy);
```
Comment (lostcontrol) [inline src/main/flight/imu.c]: sq()

File: src/main/flight/imu.c [hunk 4]
```diff
@@ -174,126 +204,222 @@ void imuCalculateAcceleration(uint32_t deltaT)
...
+    qa = q0;
+    qb = q1;
+    qc = q2;
+    q0 += (-qb * gx - qc * gy - q3 * gz);
+    q1 += (qa * gx + qc * gz - q3 * gy);
+    q2 += (qa * gy - qb * gz + q3 * gx);
+    q3 += (qa * gz + qb * gy - qc * gx);
+
+    // Normalise quaternion
+    recipNorm = invSqrt(q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3);
```
Comment (lostcontrol) [inline src/main/flight/imu.c]: sq()

File: src/main/flight/imu.c [hunk 5]
```diff
@@ -174,126 +204,222 @@ void imuCalculateAcceleration(uint32_t deltaT)
...
+static bool imuIsAccelerometerHealthy(void)
+{
+    int32_t axis;
+    int32_t accMagnitude = 0;
+
+    for (axis = 0; axis < 3; axis++) {
+        accMagnitude += (int32_t)accSmooth[axis] * accSmooth[axis];
+    }
+
+    accMagnitude = accMagnitude * 100 / ((int32_t)acc_1G * acc_1G);
```
Comment (lostcontrol) [inline src/main/flight/imu.c]: sq()

**Comment by lostcontrol** [comment]
Sorry for these minor review comments. I can't contribute more useful things for now ;-)

**Comment by digitalentity** [comment]
@lostcontrol 
No problem. After I test the latest commits I'll rebase and squash.

**Comment by digitalentity** [comment]
@hydra, please don't merge this yet. I've found some bugs and will fix them in a few days

**Comment by borisbstyle** [comment]
@digitalentity 
What bug?

**Comment by digitalentity** [comment]
GPS-related. For airplanes heading adjustment from GPS is calculated
incorrectly. Initial heading in that case is also set very slowly. I have a
fix already.

**Comment by digitalentity** [comment]
@hydra 
Bug fixed, commits squashed. You can safely merge this now.

**Comment by lostcontrol** [comment]
@digitalentity Could you provide some info about *imu_dcm_kp* and *imu_dcm_ki* please? Any needs to tune them? What are the effects?

**Comment by digitalentity** [comment]
*imu_dcm_kp* is proportional gain for gyro drift compensation. It defines weight of accelerometer/magnetometer in calculating attitude.
*imu_dcm_ki* is integral gain for gyro drift estimation, gyros we use are good enough so in general this can be kept at zero.
However, when gyro temperature change during flight (i.e. flying high altitudes where temperature is significatly lower) gyro drift can increase and using only Kp will get estimated attitude *close, but not identical* to actual attitude. If this happens, tweaking the Ki gain may help. APM uses Ki gain of 0.0087 by default, this equals to *imu_dcm_ki=87*.

**Comment by digitalentity** [comment]
In general there should be no need to change the defaults. If you have to change the defaults to get the attitude right, you are likely having issues with high vibration levels and should check your hardware.

**Comment by lostcontrol** [comment]
@digitalentity Thank you for the details. *Ki* sounds like an interesting option to tweak for winter time like now when the quad goes from 23°C to 0°C within a few minutes. Or probably better keep it outside for some minutes before flighting. Looking forward to test this PR.

**Comment by hydra** [comment]
@digitalentity I will be looking at this later this week after my move to france.  packing right now...

**Comment by digitalentity** [comment]
@lostcontrol 
I'd rather let the quad cool down before powering it up. At least this approach works for me at -15 deg :laughing: 

@hydra 
Cool! Have a good journey! :airplane: :+1:

**Comment by ledvinap** [comment]
@digitalentity: It is reasonable to expect that quad will get warmer during flight. The difference won't be that big, but it may be noticeable ...

**Comment by digitalentity** [comment]
@ledvinap 
True, that's why I didn't remove the integrator feedback from the logic. We can always dial Ki a little if it will improve things.

**Comment by hydra** [comment]
@digitalentity Moved to France ok. Still unpacking and sorting things out.  Got some very bad wifi internet.  Need to get my ADSL sorted asap.

**Comment by digitalentity** [comment]
@hydra Congrats! Hope to get your internet trouble sorted out soon!

**Comment by hydra** [comment]
@digitalentity I gave this a fly tonight on my CJMCU nano quad with a looptime of 1000 and it all seemed very stable.

I noticed after reviewing the commits that this is often repeated:

`... = DECIDEGREES_TO_DEGREES(attitude.values.yaw)`

it makes me wonder if we should store yaw in degrees in addition to decidegrees after it has been calculated.  Not suggesting we need that before a merge, just that it might reduce code size a little?

**Comment by digitalentity** [comment]
Seems reasonable. I was trying to unify the units of roll pitch and yaw.
I'll try and see if it makes code smaller.

**Comment by digitalentity** [comment]
Cool, thanks!

**Pull request merged**